### PR TITLE
Update: added ignoreExpressions option to max-classes-per-file

### LINKED ISSUE #14855: Allow excluding class expressions in max-classes-per-file
**What rule do you want to change?**
Add an option to max-classes-per-file to no longer count class expressions.

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option. Say, "allowClassExpressions"?

**Please provide some example code that this change will affect:**

```js
class VisitorFactory {
    forDescriptor(descriptor) {
        return class {
            visit(node) {
                return `Visiting ${descriptor}.`;
            }
        };
    }
}

```

**What does the rule currently do for this code?**
Counts both class declarations and expressions no matter what.

**What will the rule do after it's changed?**
If this new option is enabled, only consider class declarations.

**Are you willing to submit a pull request to implement this change?**
Possibly, but this was [proposed earlier](https://github.com/eslint/eslint/issues/11883) and the OP @JoshuaKGoldberg offered to submit a PR for this.

All of this text was copied from the [original issue](https://github.com/eslint/eslint/issues/11883). There was a [discussion](https://github.com/eslint/eslint/discussions/14684) about this and @nzakas suggested I open a new issue.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Allowed `max-classes` to take in an object for its options, with a new `ignoreExpressions` option alongside the existing `maxClasses`.

Fixes #14855.

#### Is there anything you'd like reviewers to focus on?

These files are a few years old. Would you like me to update them to a newer format?

### CODE REVIEW

File: docs/rules/max-classes-per-file.md [hunk 1]
```diff
@@ -28,8 +28,12 @@ class Foo {}
...
 ## Options
 
-This rule has a numeric option (defaulted to 1) to specify the
-maximum number of classes.
+This rule may take in either an object or a number.
+
+If the option is an object, it may contain one or both of:
+
+-   `ignoreExpressions`: a boolean option (defaulted to `false`) to ignore class expressions.
+-   `maximum`: a numeric option (defaulted to 1) to specify the maximum number of classes.
```
Comment (nzakas) [inline docs/rules/max-classes-per-file.md:36]: We generally use “max” for this purpose in other rules, so good to be consistent. 

```suggestion
-   `max`: a numeric option (defaulted to 1) to specify the maximum number of classes.
```

File: docs/rules/max-classes-per-file.md [hunk 2]
```diff
@@ -28,8 +28,12 @@ class Foo {}
 
 ## Options
 
-This rule has a numeric option (defaulted to 1) to specify the
-maximum number of classes.
+This rule may take in either an object or a number.
```
Comment (nzakas) [inline docs/rules/max-classes-per-file.md:31]:
```suggestion
This rule may be configured with either an object or a number.
```

File: docs/rules/max-classes-per-file.md [hunk 3]
```diff
@@ -39,11 +43,36 @@ For example:
 }
 ```
 
-Examples of **correct** code for this rule with the numeric option set to `2`:
+```json
+{
+    "max-classes-per-file": [
+        "error",
+        { "ignoreExpressions": true, "maximum": 2 }
```
Comment (nzakas) [inline docs/rules/max-classes-per-file.md:50]:
```suggestion
        { "ignoreExpressions": true, "max": 2 }
```

File: docs/rules/max-classes-per-file.md [hunk 4]
```diff
@@ -39,11 +43,36 @@ For example:
...
+    ]
+}
+```
+
+Examples of **correct** code for this rule with the `maximum` option set to `2`:
```
Comment (nzakas) [inline docs/rules/max-classes-per-file.md:55]:
```suggestion
Examples of **correct** code for this rule with the `max` option set to `2`:
```

File: lib/rules/max-classes-per-file.js [hunk 5]
```diff
@@ -25,8 +25,25 @@ module.exports = {
...
+                        type: "integer",
+                        minimum: 1
+                    },
+                    {
+                        type: "object",
+                        properties: {
+                            ignoreExpressions: {
+                                type: "boolean"
+                            },
+                            maxClasses: {
```
Comment (nzakas) [inline lib/rules/max-classes-per-file.js:39]:
```suggestion
                            max: {
```

File: lib/rules/max-classes-per-file.js [hunk 6]
```diff
@@ -35,8 +52,10 @@ module.exports = {
         }
     },
     create(context) {
-
-        const maxClasses = context.options[0] || 1;
+        const [option = {}] = context.options;
+        const [ignoreExpressions, maxClasses] = typeof option === "number"
```
Comment (nzakas) [inline lib/rules/max-classes-per-file.js:56]:
```suggestion
        const [ignoreExpressions, max] = typeof option === "number"
```

File: lib/rules/max-classes-per-file.js [hunk 7]
```diff
@@ -35,8 +52,10 @@ module.exports = {
...
+            ? [false, option || 1]
+            : [option.ignoreExpressions, option.maxClasses || 1];
```
Comment (nzakas) [inline lib/rules/max-classes-per-file.js:58]:
```suggestion
            : [option.ignoreExpressions, option.max || 1];
```

File: tests/lib/rules/max-classes-per-file.js [hunk 8]
```diff
@@ -29,6 +29,29 @@ ruleTester.run("max-classes-per-file", rule, {
         {
             code: "class Foo {}\nclass Bar {}",
             options: [2]
+        },
+        {
+            code: "class Foo {}",
+            options: [{ maxClasses: 1 }]
```
Comment (nzakas) [inline tests/lib/rules/max-classes-per-file.js:35]:
```suggestion
            options: [{ max: 1 }]
```

File: tests/lib/rules/max-classes-per-file.js [hunk 9]
```diff
@@ -29,6 +29,29 @@ ruleTester.run("max-classes-per-file", rule, {
...
+        },
+        {
+            code: "class Foo {}\nclass Bar {}",
+            options: [{ maxClasses: 2 }]
```
Comment (nzakas) [inline tests/lib/rules/max-classes-per-file.js:39]:
```suggestion
            options: [{ max: 2 }]
```

File: tests/lib/rules/max-classes-per-file.js [hunk 10]
```diff
@@ -29,6 +29,29 @@ ruleTester.run("max-classes-per-file", rule, {
...
+        },
+        {
+            code: `
+                class Foo {}
+                const myExpression = class {}
+            `,
+            options: [{ ignoreExpressions: true, maxClasses: 1 }]
```
Comment (nzakas) [inline tests/lib/rules/max-classes-per-file.js:46]:
```suggestion
            options: [{ ignoreExpressions: true, max: 1 }]
```

File: tests/lib/rules/max-classes-per-file.js [hunk 11]
```diff
@@ -29,6 +29,29 @@ ruleTester.run("max-classes-per-file", rule, {
...
+        },
+        {
+            code: `
+                class Foo {}
+                class Bar {}
+                const myExpression = class {}
+            `,
+            options: [{ ignoreExpressions: true, maxClasses: 2 }]
```
Comment (nzakas) [inline tests/lib/rules/max-classes-per-file.js:54]:
```suggestion
            options: [{ ignoreExpressions: true, max: 2 }]
```

File: tests/lib/rules/max-classes-per-file.js [hunk 12]
```diff
@@ -54,6 +81,25 @@ ruleTester.run("max-classes-per-file", rule, {
...
             options: [2],
             errors: [{ messageId: "maximumExceeded", type: "Program" }]
+        },
+        {
+            code: `
+                class Foo {}
+                class Bar {}
+                const myExpression = class {}
+            `,
+            options: [{ ignoreExpressions: true, maxClasses: 1 }],
```
Comment (nzakas) [inline tests/lib/rules/max-classes-per-file.js:91]:
```suggestion
            options: [{ ignoreExpressions: true, max: 1 }],
```

File: tests/lib/rules/max-classes-per-file.js [hunk 13]
```diff
@@ -54,6 +81,25 @@ ruleTester.run("max-classes-per-file", rule, {
...
+            errors: [{ messageId: "maximumExceeded", type: "Program" }]
+        },
+        {
+            code: `
+                class Foo {}
+                class Bar {}
+                class Baz {}
+                const myExpression = class {}
+            `,
+            options: [{ ignoreExpressions: true, maxClasses: 2 }],
```
Comment (nzakas) [inline tests/lib/rules/max-classes-per-file.js:101]:
```suggestion
            options: [{ ignoreExpressions: true, max: 2 }],
```

**Review by nzakas — CHANGES_REQUESTED** [review]
The docs specified the option name as “maximum” while the code used “maxClasses”. I think we can just go with “max”, which is what other rules use in a similar situation.

code looks good overall.

**Review by nzakas — APPROVED** [review]
LGTM. Thanks!

**Review by snitin315 — APPROVED** [review]
LGTM, thanks for contributing!

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**